### PR TITLE
kvserver: make Raft ticks adaptive to delays

### LINKED ISSUE #66686: kv: raft ticks are not adaptive to slow scheduling
Related to https://github.com/cockroachdb/cockroach/issues/66682.
Related to https://github.com/cockroachdb/cockroach/issues/17500.

We configure Raft to tick each Raft group every 200ms (`defaultRaftTickInterval`). We also configure leaders to send Raft heartbeats every 5 ticks (`defaultRaftHeartbeatIntervalTicks`). Finally, we configure followers to call elections every 15 ticks (`defaultRaftElectionTimeoutTicks`). In the happy case, this all works.

However, there are various reasons why a raft tick may not be scheduled for over 200ms. The obvious case is scheduler overload. But the case that we see more commonly is a single Range being slow to either process incoming requests [here](https://github.com/cockroachdb/cockroach/blob/e1697d1728be9165adad55b489676773062d4ef4/pkg/kv/kvserver/scheduler.go#L272) or perform synchronous local I/O [here](https://github.com/cockroachdb/cockroach/blob/e1697d1728be9165adad55b489676773062d4ef4/pkg/kv/kvserver/scheduler.go#L284). This can delay a Raft tick for more than 200ms, which can delay heartbeats, which can cause leadership instability.

If we are starving all Raft progress for 3+ seconds, we certainly won't be able to hold stable leadership. However, there's a less severe case that can be just as devastating and is likely more common. A leader will only send out a heartbeat every 5 ticks. So even if a leader is delayed in processing ticks by only 600ms, it will not send out heartbeats fast enough to maintain leadership. This is what we saw in #66682.

We should be able to provide some stability here by making raft ticks more adaptive to scheduling delays. For instance, if we want to tick every 200ms but we are delayed by 600ms between ticks, each call to [`replica.tick`](https://github.com/cockroachdb/cockroach/blob/d0cdd738c54df6ea8a73cc06a92bd5812947605d/pkg/kv/kvserver/replica_raft.go#L956) could call Tick on the internal raft group 3 times. This would ensure that heartbeats are still sent out roughly every second. This would allow us to make a rough guarantee that as long as a leader is scheduled by the Raft scheduler once every 3 seconds, it will be able to retain leadership (if we want it to).

/cc @cockroachdb/kv

Jira issue: CRDB-8183

### PR DESCRIPTION
**Pull request by pav-kv**
Previously, a relatively short Raft tick delay (600ms instead of 200ms in
default configuration) persisting over a few ticks could cause a replica to
lose leadership, or (if follower) become a candidate too slowly.

If processing a queued tick is delayed by more than one tick's real time
equivalent (200ms), chances are that the next tick is a no-op (as the scheduler
does not enqueue it while there is another tick in the queue), and we won't let
the Raft package know that more time has passed. This creates a skew between
ticks and real time, i.e. the average real time duration per tick in the
relevant window of O(election timeout) is too low. It seems important to
minimize this skew to avoid unnecessary election events.

This commit makes Raft ticks adaptive to scheduling delays. In addition to
remembering 1 bit of whether a tick should occur or not, the scheduler now
remembers the number of ticks that has to happen. The worker induces this many
ticks when it gets to process this Raft group.

Fixes #66686

Release note (performance improvement): Raft ticks now adapt to scheduling
delays. This helps preventing re-elections, and the corresponding performance
effects, in the event of relatively short (sub-second) processing delays.

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
TODO in this PR:

- [ ] Test: reproduce the time warp on a leader.
- [ ] Test: reproduce it on a follower.

**Comment by pavelkalinnikov** [comment]
Alternatives considered:

1. Re-use the already computed scheduling [latency](https://github.com/cockroachdb/cockroach/blob/f292c7a6f136af0b4f1cb9ee8de7af6c1107ac08/pkg/kv/kvserver/scheduler.go#L286-L287), e.g. divide it by tick duration and round up to get the number of ticks to induce.

Pros: simple. Cons:

- latency is computed before `processRequestQueue` which is one of the things that can be slow, so the actual pre-tick latency is different,
- the sub-tick calculation error keeps accumulating over time,
- scheduler is currently fully driven by ticks, and is easy to test. Relying on time here will make it mixed, and more subtle to test.

2. Send "loopback" ticks to the queue. Cons: The queue is batched, sending single entries will make it slower. It can make the situation of delayed scheduling even worse.

File: pkg/kv/kvserver/scheduler.go [hunk 1]
```diff
@@ -297,7 +301,7 @@ func (s *raftScheduler) worker(ctx context.Context) {
 				state.flags |= stateRaftReady
 			}
 		}
-		if state.flags&stateRaftTick != 0 {
+		for t := state.ticks; t > 0; t-- {
 			// processRaftTick returns true if the range should perform ready
 			// processing. Do not reorder this below the call to processReady.
 			if s.processor.processTick(ctx, id) {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/scheduler.go:307]: The call stack here is:

1. `Store.processTick`
2. `Replica.tick`
3. `raft.RawNode.Tick`

We can loop over one of them (currently 1-st). Seemingly, it's more efficient to do deeper in the stack, but I wonder which one is correct. Perhaps we can refactor to enable looping over `RawNode.Tick` directly.

Comment (erikgrinaker) [reply pkg/kv/kvserver/scheduler.go:307]: Yeah, I'd loop over `processTick`. There may be logic that high up in the stack that relies (or will rely on) the tick count for correctness.

File: pkg/kv/kvserver/scheduler.go [hunk 2]
```diff
@@ -344,13 +348,19 @@ func (s *raftScheduler) worker(ctx context.Context) {
...
 	prevState := s.mu.state[id]
-	if prevState.flags&addFlags == addFlags {
+	if prevState.flags&addFlags == addFlags && !ticked {
 		return 0
 	}
 	var queued int
 	newState := prevState
 	newState.flags = newState.flags | addFlags
+	if ticked {
+		newState.ticks++ // TODO(pavelkalinnikov): Check overflows, or have a limit.
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/scheduler.go:362]: I wouldn't worry about overflow. If we're ticking at 200 ms, it'll overflow in 10 billion years.

File: pkg/kv/kvserver/scheduler.go [hunk 3]
```diff
@@ -297,7 +301,7 @@ func (s *raftScheduler) worker(ctx context.Context) {
 				state.flags |= stateRaftReady
 			}
 		}
-		if state.flags&stateRaftTick != 0 {
+		for t := state.ticks; t > 0; t-- {
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/scheduler.go:304]: We should have an assertion somewhere that `raftStateTick` also has `ticks > 0`, or guard against it.

File: pkg/kv/kvserver/scheduler.go [hunk 4]
```diff
@@ -344,13 +348,19 @@ func (s *raftScheduler) worker(ctx context.Context) {
 func (s *raftScheduler) enqueue1Locked(
 	addFlags raftScheduleFlags, id roachpb.RangeID, now int64,
 ) int {
+	// TODO(pavelkalinnikov): Pass ticks increment to avoid branching below.
+	ticked := addFlags&stateRaftTick != 0
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/scheduler.go:352]: We can make it branchless by dividing by `stateRaftTick`, which will yield 1, which can be added to `newState.ticks`.

**Comment by erikgrinaker** [comment]
>Suggested solution pros: accurate, stays driven by external ticks.

I like this solution, keeps wall time out of it. Let's go with this.

File: pkg/kv/kvserver/scheduler.go (hunk 2 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/scheduler.go:362]: Good point. It's 13.6 years though (if I'm not mistaken that max `int` is 2^31-1). It still seems enough, but a bit on the edge (esp. if tick time is tuned down to order of milliseconds).

File: pkg/kv/kvserver/scheduler.go (hunk 3 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/scheduler.go:304]: I wonder if we should just remove `raftStateTick` because it's redundant. You can see in this PR that it's effectively ignored in all the computation, we could just leave `ticks` instead of it.

File: pkg/kv/kvserver/scheduler.go (hunk 2 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/scheduler.go:362]: `int` defaults to 64-bit on 64-bit architectures. I think that's the only one we really care about, not sure if we even support 32-bit.

File: pkg/kv/kvserver/scheduler.go (hunk 3 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/scheduler.go:304]: It is redundant, but I think I'd still want to leave it in -- the state flags should cover all states, I think.

**Comment by pavelkalinnikov** [comment]
@erikgrinaker @tbg PTAL

File: pkg/kv/kvserver/scheduler.go [hunk 5]
```diff
@@ -310,7 +316,7 @@ func (s *raftScheduler) worker(ctx context.Context) {
 
 		s.mu.Lock()
 		state = s.mu.state[id]
-		if state.flags == stateQueued {
+		if state.flags == stateQueued && state.ticks == 0 {
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/scheduler.go:319]: We shouldn't need this check because we also set `stateRaftTick`.

File: pkg/kv/kvserver/scheduler.go [hunk 6]
```diff
@@ -297,7 +303,7 @@ func (s *raftScheduler) worker(ctx context.Context) {
 				state.flags |= stateRaftReady
 			}
 		}
-		if state.flags&stateRaftTick != 0 {
+		for tick, t := state.flags&stateRaftTick != 0, state.ticks; tick && t > 0; t-- {
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/scheduler.go:306]: nit: may want to just have an if wrapping the for-loop here. While I appreciate the terseness, it's a bit hard to parse.

File: pkg/kv/kvserver/scheduler.go (hunk 3 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/scheduler.go]: Thanks! Let's also add an assertion in `worker()`. We can gate it on `RaceEnabled` to avoid any unintended production crashes and overhead (not ideal, but it's a convention we sometimes use).

```
if util.RaceEnabled && (state.flags&stateRaftTick != 0 && state.ticks == 0) || (state.flags&stateRaftTick == 0 && state.ticks != 0) {
    log.Fatalf("stateRaftTick %v with ticks %v", state.flags&stateRaftTick, state.ticks)
}
```

File: pkg/kv/kvserver/scheduler_test.go [hunk 7]
```diff
@@ -273,8 +275,8 @@ func TestSchedulerBuffering(t *testing.T) {
 	}{
 		{stateRaftReady, "ready=[1:1] request=[] tick=[]"},
 		{stateRaftRequest, "ready=[1:1] request=[1:1] tick=[]"},
-		{stateRaftTick, "ready=[1:1] request=[1:1] tick=[1:1]"},
-		{stateRaftReady | stateRaftRequest | stateRaftTick, "ready=[1:2] request=[1:2] tick=[1:2]"},
+		{stateRaftTick, "ready=[1:1] request=[1:1] tick=[1:5]"},
+		{stateRaftReady | stateRaftRequest | stateRaftTick, "ready=[1:2] request=[1:2] tick=[1:10]"},
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/scheduler_test.go:279]: Nice, this turned out to be straightforward.

**Review by erikgrinaker — APPROVED** [review]
LGTM. I'm leaning towards letting this sit until after we cut the release branch. Wdyt @tbg, too conservative?

**Comment by tbg** [comment]
Yes let's let it sit.

One thought I had is that it might be good to have a max number of ticks we are going to buffer. Consider the case in which we have a group of replicas that is "slow" and others are "fast". The slow ones will accumulate ticks, meaning that when they get ticked they will keep a scheduler worker busy for far longer. With enough slow ranges, this will starve the scheduler and won't be serving the "fast" ranges.
In other words, the PR as is says "if you deserve a tick you will get do carry out the tick, no matter how long it takes". But this has runaway potential if processing a single tick takes longer than the tick interval (the next time around you'll do two ticks, which takes ~twice as long, so next time you get three ticks, etc...). So maybe we'll set a conservative cap here.

**Review by tbg — APPROVED** [review]

**Review by pav-kv — COMMENTED** [review]
@erikgrinaker Made some changes and incorporated the idea of limiting the number of ticks from @tbg. PTAL.

File: pkg/kv/kvserver/scheduler.go [hunk 8]
```diff
@@ -26,6 +26,7 @@ import (
 )
 
 const rangeIDChunkSize = 1000
+const maxSchedulerTicks = 10
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/scheduler.go:29]: Let's use `RaftConfig.RaftElectionTimeoutTicks` for this if we have easy access to it, or otherwise set it to `defaultRaftElectionTimeoutTicks` (15). This would avoid delaying elections.

**Review by erikgrinaker — APPROVED** [review]

File: pkg/kv/kvserver/scheduler.go (hunk 8 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/scheduler.go]: Good idea. Done.

**Comment by pavelkalinnikov** [comment]
Fixing the test.

File: pkg/kv/kvserver/scheduler.go (hunk 3 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/scheduler.go]: Done, with a bit of nesting for better readability.

**Comment by pavelkalinnikov** [comment]
bors r=erikgrinaker,tbg
